Testing/test_rate.py

```diff
@@ -17,77 +14,59 @@
 
 
 class TestRate(unittest.TestCase):
-    def setUp(self):
-        # agents
-
-        self.a1 = AtomicAgent("T", "i")
-        self.a2 = AtomicAgent("S", "i")
-        self.a3 = AtomicAgent("T", "a")
-        self.a4 = AtomicAgent("S", "a")
-
-        self.s1 = StructureAgent("X", {self.a4})
-        self.s2 = StructureAgent("X", {self.a2})
-        self.s3 = StructureAgent("K", {self.a3})
-        self.s4 = StructureAgent("K", {self.a1})
-        self.s5 = StructureAgent("X", set())
-        self.s6 = StructureAgent("K", set())
-
-        self.c1 = Complex([self.s6], "cyt")  # K()::cyt
-        self.c2 = Complex([self.s3], "cyt")  # K(T{a})::cyt
-        self.c3 = Complex([self.s4], "cyt")  # K(T{i})::cyt
-        self.c4 = Complex([self.s4, self.s1], "cyt")  # K(T{i}).X(S{a})::cyt
-        self.c5 = Complex([self.s4, self.s2], "cyt")  # K(T{i}).X(S{i})::cyt
-        self.c6 = Complex([self.s3, self.s1], "cyt")  # K(T{a}).X(S{a})::cyt
-        self.c7 = Complex([self.s3, self.s2], "cyt")  # K(T{a}).X(S{i})::cyt
-
-        # rates
-
-        self.parser = Parser("rate")
-        rate_expr = "3.0*[K()::cyt]/2.0*v_1"
-
-        self.rate_1 = Rate(self.parser.parse(rate_expr).data)
-
-        rate_expr = "3.0*[K(T{i}).X()::cyt] + [K()::cyt]"
-
-        self.rate_2 = Rate(self.parser.parse(rate_expr).data)
-
-        rate_expr = "(3.0*[K()::cyt])/(2.0*v_1)"
-
-        self.rate_3 = Rate(self.parser.parse(rate_expr).data)
-
-        # states
-
-        self.state_1 = State(Vector(np.array([2, 3])), Memory(0))
-        self.state_2 = State(Vector(np.array([2, 0, 3, 1, 6, 2])), Memory(0))
-
     def test_to_str(self):
-        self.assertEqual(str(self.rate_1), "3.0*[K()::cyt]/2.0*v_1")
+        self.assertEqual(str(objects.rate1), "3.0*[K()::cyt]/2.0*v_1")
 
     def test_eq(self):
-        self.assertEqual(self.rate_2, self.rate_2)
-        self.assertNotEqual(self.rate_1, self.rate_2)
+        self.assertEqual(objects.rate2, objects.rate2)
+        self.assertNotEqual(objects.rate1, objects.rate2)
 
     def test_vectorize(self):
-        ordering = (self.c2, self.c3)
-        self.assertEqual(self.rate_1.vectorize(ordering, dict()), [Vector(np.array([1, 1]))])
-        ordering = (self.c2, self.c3, self.c4, self.c5, self.c6, self.c7)
-        self.assertEqual(self.rate_2.vectorize(ordering, dict()),
-                         [Vector(np.array([0, 0, 1, 1, 0, 0])), Vector(np.array([1, 1, 0, 0, 0, 0]))])
+        ordering = (objects.c15, objects.c16)
+        self.assertEqual(
+            objects.rate1.vectorize(ordering, dict()), [Vector(np.array([1, 1]))]
```

Comment:
There is a test failing here (and also in `test_evaluate`).

```
lark.exceptions.VisitError: Error trying to process rule "agent":
E               
E               'Vector' object has no attribute 'compatible'
```

It is not clear what is the reason. It looks like the agent inside `ordering` is already `vectorized`. I hope this does not implicate that some other test did that and internally it stayed like that. Python can be sometimes weird with mutable objects. Try to investigate (e.g. duplicate the used objects where defined and used them only here if it helps).

Reply:
This problem should be solved in the commit 8af17f2. I did not find the exact cause, but not creating the rate objects in objects_testing.py solved the issue.